### PR TITLE
Use gross roof area for ave ceiling height

### LINKED ISSUE #8770: Use gross ceiling surface area to calculate average ceiling height
Issue overview
--------------
In the surface geometry, the ceiling height of a zone is calculated using the averaged ceiling height for all the ceiling (roof surfaces. For the case of multiple surfaces, a weight to each ceiling surface is given by the fraction of the individual ceiling surface area among the summed area of all the ceiling surfaces. 

However, the area used for this weighting is the “net” surface area—rather than the gross one. In the case of a roof surface contains a window (a skylight, for example), the corresponding weight of that surface is then would be reduced. This will make the calculated results to be skewed toward the roof that has no windows or less window areas. 

Since the “area” of the ceiling in this context is just a geometric concept that does need to consider how much of window surface area it contains, it is suggested that the ceiling gross areas rather than the (net) area should be to be used for this calculation. 

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): All
 - Version of EnergyPlus (if using an intermediate build, include SHA): v9.5 and before, probably all
 - Related Helpdesk ticket: 15707 (related but not exactly the same issue) 

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added EnergyPlusDevSupport\DefectFiles
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
 - Fixes #8770 
 - 

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mitchute** [comment]
@jcyuan2020 can you pull in develop one more time so I can do a final review and merge?

**Comment by jcyuan2020** [comment]
> 
> 
> @jcyuan2020 can you pull in develop one more time so I can do a final review and merge?
@mitchute Thanks! The branch is now up-to-date.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 1]
```diff
@@ -7021,3 +7021,532 @@ TEST_F(EnergyPlusFixture, GetSurfaceData_SurfaceOrder)
...
+        "    4,                       !- Number of Vertices",
+        "    2, 4, 4,                         !- X,Y,Z  1 {m}",
+        "    2, 0, 4,                         !- X,Y,Z  2 {m}",
+        "    6, 0, 4,                         !- X,Y,Z  3 {m}",
+        "    6, 4, 4;                         !- X,Y,Z  4 {m}",
+
+        "BuildingSurface:Detailed,",
+        "06Roof_E,                !- Name",
+        "Roof,                    !- Surface Type",
+        "R13WALL,                 !- Construction Name",
```
Comment (mitchute) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:7457]: Is this a mistake? Surface name "06Roof_E" and Construction Name "R13Wall"?

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 2]
```diff
@@ -7021,3 +7021,532 @@ TEST_F(EnergyPlusFixture, GetSurfaceData_SurfaceOrder)
...
+        "4,                       !- Number of Vertices",
+        "6, 0, 4,                         !- X,Y,Z  1 {m}",
+        "8, 0, 2.5,                          !- X,Y,Z  2 {m}",
+        "8, 4, 2.5,                          !- X,Y,Z  3 {m}",
+        "6, 4, 4;                         !- X,Y,Z  4 {m}",
+
+        "BuildingSurface:Detailed,",
+        "07Roof_W,                !- Name",
+        "Roof,                    !- Surface Type",
+        "R13WALL,                 !- Construction Name",
```
Comment (mitchute) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:7473]: Same here.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 3]
```diff
@@ -7021,3 +7021,532 @@ TEST_F(EnergyPlusFixture, GetSurfaceData_SurfaceOrder)
...
+    Real64 totalGrossCeilingArea = 0.0;
+    totalGrossCeilingArea =
+        state->dataSurface->Surface(6).GrossArea + state->dataSurface->Surface(7).GrossArea + state->dataSurface->Surface(8).GrossArea;
+
+    Real64 totalNetCeilingArea = 0.0;
+    totalNetCeilingArea = state->dataSurface->Surface(6).Area + state->dataSurface->Surface(7).Area + state->dataSurface->Surface(8).Area;
+
+    Real64 ceilingHeight_expected = 0.0;
+    ceilingHeight_expected = 3.25 * (state->dataSurface->Surface(7).GrossArea + state->dataSurface->Surface(8).GrossArea) / totalGrossCeilingArea +
+                             4.0 * state->dataSurface->Surface(6).GrossArea / totalGrossCeilingArea;
```
Comment (mitchute) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:7544]: What do the "3.25" and "4.0" signify here?

**Review by mitchute — COMMENTED** [review]
This fix seems appropriate based on the issue description. There were a few minor comments to address; however, one larger question is why there are no math diffs here? Are there any example files with zones containing multiple ceiling sub-surfaces where this would apply?

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc (hunk 1 above)
Comment (jcyuan2020) [reply tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:7457]: No, the construction actually does not matter here as far it is a normal construction that would be counted to gross or net area of a surface. Here only a simple (or simplest) version.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc (hunk 2 above)
Comment (jcyuan2020) [reply tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:7473]: Same as above--construction does not matter here.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc (hunk 3 above)
Comment (jcyuan2020) [reply tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:7544]: 3.25 and 4.0 (in meters) are the average height of the respective roof surfaces. This is an example of multiple roof surfaces in the same zone, with different heights--two ceiling surfaces has height of 3.25 and one has a height of 4m in the unit test case.

**Comment by jcyuan2020** [comment]
> 
> 
> This fix seems appropriate based on the issue description. There were a few minor comments to address; however, one larger question is why there are no math diffs here? Are there any example files with zones containing multiple ceiling sub-surfaces where this would apply?

@mitchute Thanks for the review and the comments! There are two special conditions simultaneously required for the diff to show up: the first one is that a zone that contains two or more roof surfaces at different heights; and second one is that the ceiling surfaces contain a window (or skylight)--at least one of them needs to have some window (which make the net surface area not equal to the gross surface area. 

The above two conditions rarely happens (or do not happen at all) in the test file examples. But the situation is not difficult to imagine or encounter in reality (may not be a rare case). Actually it could be a commonly seen scenario, such as in residential house wit asymmetric pitched roofs where the two roofs surfaces could start at different height and also one side may have a skylight (or window). In helpdesk ticket 15707 the user also modeled such a case (consider this is an actual simulation configuration that reflected a real case from a modeler). The defect file of the issue post also shows such an example (it is the same three-roof/ceiling surfaces example in the test file).

**Review by mitchute — APPROVED** [review]

**Comment by mitchute** [comment]
OK, thanks for the explanation. Unit tests all pass locally. Merging.

**Pull request merged**